File: serializeTree.py
import json


class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None


def serialize(root):
    if root == None:
        return None

    serialize_tree = dict()
    left = serialize(root.left)
    right = serialize(root.right)

    serialize_tree['data'] = root.data
    if left:
        serialize_tree['left'] = left
    if right:
        serialize_tree['right'] = right

    return json.dumps(serialize_tree)


def deserialize(data):
    serialize_tree = json.loads(data)

    node = Node(serialize_tree['data'])
    if 'left' in serialize_tree:
        node.left = deserialize(serialize_tree['left'])
    if 'right' in serialize_tree:
        node.right = deserialize(serialize_tree['right'])

    return node


node_a = Node('a')
node_b = Node('b')
node_c = Node('c')
node_d = Node('d')
node_e = Node('e')
node_f = Node('f')
node_g = Node('g')
node_a.left = node_b
node_a.right = node_c
node_b.left = node_d
node_b.right = node_e
node_c.left = node_f
node_c.right = node_g

data = serialize(node_a)
print(data)
copy = deserialize(data)

File: serializeTree.cpp
#include <bits/stdc++.h>

using namespace std;
using ll = long long;

struct node
{
    int data;
    struct node *left;
    struct node *right;

    node(int val)
    {
        data = val;
        left = nullptr;
        right = nullptr;
    }
};

int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(0);

    struct node *root = new node(10);
    root->left = new node(11);
    root->right = new node(20);
    root->left->left = new node(23);
    root->left->right = new node(8);
    root->right->left = new node(111);
    root->right->right = new node(33);

    return 0;
}
